persistent-postgresql/test/Streaming.hs

```diff
@@ -0,0 +1,58 @@
+{-# LANGUAGE DataKinds #-}
+{-# LANGUAGE DerivingStrategies #-}
+{-# LANGUAGE ExistentialQuantification #-}
+{-# LANGUAGE FlexibleInstances #-}
+{-# LANGUAGE GeneralizedNewtypeDeriving #-}
+{-# LANGUAGE MultiParamTypeClasses #-}
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE QuasiQuotes #-}
+{-# LANGUAGE StandaloneDeriving #-}
+{-# LANGUAGE TemplateHaskell #-}
+{-# LANGUAGE TypeApplications #-}
+{-# LANGUAGE TypeFamilies #-}
+{-# LANGUAGE UndecidableInstances #-}
+
+module Streaming where
+
+import PgInit
+
+import Data.Conduit
+import qualified Data.Conduit.Combinators as CC
+import qualified Data.Conduit.List as CL
+
+share [mkPersist sqlSettings, mkMigrate "streamingMigrate"] [persistLowerCase|
+
+StreamableNumber
+    val Int
+    deriving Eq Show Ord
+
+|]
+
+numberOfRows :: Int
+numberOfRows = 1000000
+
+numbers :: Monad m => ConduitT () StreamableNumber m ()
+numbers = CC.unfold (\n -> if n >= numberOfRows then Nothing else Just (StreamableNumber n, n + 1)) 1
+
+wipe :: IO ()
+wipe = runConnAssert $ do
+    deleteWhere ([] :: [Filter StreamableNumber])
+
+itDb :: String -> SqlPersistT (LoggingT (ResourceT IO)) a -> SpecWith (Arg (IO ()))
+itDb msg action = it msg $ runConnAssert $ void action
+
+specs :: Spec
+specs = describe "Streaming rows" $ do
+    describe "selectStream" $ before_ wipe $ do
+        itDb "streams rows" $ do
+          runConduit $ numbers .| CL.chunksOf 1000 .| CC.mapM_ (void . insertMany)
+
+          let resultStream = selectStream [] [ Asc StreamableNumberVal ]
+
+          let checkIncrementing expected (Entity _ (StreamableNumber actual)) = do
+                expected `shouldBe` actual
+                pure $ expected + 1
+
+          result <- runConduit $ resultStream .| CC.foldM checkIncrementing 1
+
+          result `shouldBe` numberOfRows
```

Comment:
I'll double check it. I believe the nature of cursor queries should automatically avoid some of these concurrency issues, since streaming really fetches batches of rows from the cursor by name.

It might be necessary to make cursor statements bypass the `connStmtMap`, since the randomly-generated cursor name isn't included in the prepared query.

Reply:
I'm struggling to follow the exact control flow, but: I believe this implementation is actually immune to some of these concurrency issues.

Testing has shown that certainly the single-threaded version behaves as expected. Conduits which simultaneously run the same query in a single-threaded application (one of the problematic uses mentioned in #981) yield results in the expected way, and don't affect each other. I *think* this is for the reason I mentioned above - each call to `withCursorStmt'` will cause a new temporary cursor to be created, the name of which is then used to grab rows in chunks.

The multi-threaded version might well still have issues - but testing that can wait until `ersistent-postgresql-streaming` is spun off.